dockerunit-consul/src/main/java/com/github/qzagarese/dockerunit/discovery/consul/ConsulDiscoveryProvider.java

```diff
@@ -114,21 +114,45 @@ private Service doCleanup(Service current, Service global) {
 	}
 
 	private int findPort(InspectContainerResponse response, List<ServiceRecord> records) {
-		ServiceRecord record = records.stream()
-			.filter(r -> this.matchPort(r, response))
-			.findFirst()
-			.orElseThrow(() ->  
-				new RuntimeException("Cannot find exposed port/ip for container " + response.getName()));
-		return record.getPort();
+		Optional<ServiceRecord> record = records.stream()
+			.filter(r -> this.matchRecord(r, response))
+			.findFirst();
+		return record.isPresent() ? extractMappedPort(record.get().getPort(), response) : 0;
 	}
 
-	private boolean matchPort(ServiceRecord record, InspectContainerResponse r) {
-		return r.getNetworkSettings().getPorts().getBindings().values()
-                .stream()
-                .map(bindings -> Optional.ofNullable(bindings).orElse(new Binding[]{}))
-                .filter(b -> b.length > 0)
-                .map(b -> parsePort(b[0].getHostPortSpec()))
-                .anyMatch(port -> record.getPort() == port.orElse(-1));
+	private int extractMappedPort(int port, InspectContainerResponse response) {
+		return response.getNetworkSettings().getPorts().getBindings().entrySet().stream()
+				.filter(entry -> entry.getKey().getPort() == port)
+				.map(entry -> entry.getValue())
+				.filter(bindings -> bindings != null && bindings.length > 0)
+				.findFirst()
+				.map(bindings -> parsePort(bindings[0].getHostPortSpec()).orElse(0))
+				.get();
+	}
+
+	private boolean matchRecord(ServiceRecord record, InspectContainerResponse r) {
+		return matchIP(record.getServiceAddress(), r) && matchPort(record.getPort(), r);
+	}
+
+	private boolean matchPort(int port, InspectContainerResponse r) {
+		return r.getNetworkSettings().getPorts().getBindings().entrySet().stream()
+				.map(entry -> entry.getKey().getPort())
+				.filter(p -> p == port)
+				.findFirst()
+				.isPresent();
+	}
+
+	private boolean matchIP(String address, InspectContainerResponse r) {
+		return null != address && address.equals(extractIp(r));
+	}
+
+	private String extractIp(InspectContainerResponse r) {
```

Comment:
the logic in this method is repeated in `UseConsulDnsExtensionInterpreter` maybe create a "ConsulAddressFinder" ? something like:
```
Optinal<String> ipAddress = ConsulAddressFinder.forNetwork(ContainerNetwork network)
        .find("bridge");

String ipAddress = ConsulAddressFinder.forNetwork(ContainerNetwork network)
        .withDefault(DOCKER_BRIDGE_IP_PROPERTY, DOCKER_BRIDGE_IP_DEFAULT)
        .find("bridge");
```
(can be a injected singleton too instead of a static reference but you get the gist - in fact might be worth thinking about whether this "find" feature for ports and ip addresses should be extracted to a different class as it's not really consul-specific)

Reply:
It's not Consul specific.
I'll pull put a class.